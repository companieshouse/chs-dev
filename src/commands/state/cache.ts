import { Args, Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Cache extends Command {

    static description = "Cache the state of chs-dev into a saved file";

    static args = {
        name: Args.string({
            required: false,
            description: "Name of the cache"
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { args, flags } = await this.parse(Cache);
        const timestamp: string = new Date(Date.now()).toISOString();
        const cacheName: string = args.name || timestamp;

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`save chs-dev state?`);
        };

        if (await handlePrompt()) {
            //
        }
    }

}
