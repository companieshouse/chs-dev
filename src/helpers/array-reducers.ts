export const collect = <U, T>(fn: (value: T) => U[]): (result: U[], value: T) => U[] => {
    return (values: U[], value: T): U[] => {
        values.push(...fn(value));
        return values;
    };
};

export const deduplicate = (uniqueValues: string[], value: string): string[] => {
    if (!uniqueValues.includes(value)) {
        uniqueValues.push(value);
    }
    return uniqueValues;
};
