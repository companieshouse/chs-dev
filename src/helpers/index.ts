/**
 * Checks if a given value is an object and not an array.
 *
 * @param {unknown} T - The value to check.
 * @returns {boolean} - Returns `true` if the value is an object and not an array, otherwise `false`.
 */
export const isObjectFormat = (T: unknown): boolean => {
    return typeof T === "object" && !Array.isArray(T) && T !== null;
};

/**
 * Determines if an object is empty.
 *
 * @param {Record<string, any>} obj - The object to check.
 * @returns {boolean} - Returns `true` if the object has no own enumerable properties, otherwise `false`.
 */
export const isEmptyObject = (obj: Record<string, any>): boolean => {
    return Object.keys(obj).length === 0;
};

/**
 * Checks if there is at least one common element between two arrays.
 *
 * @param {unknown[]} array1 - The first array to compare.
 * @param {unknown[]} array2 - The second array to compare.
 * @returns {boolean} - Returns `true` if any element in `array1` exists in `array2`, otherwise `false`.
 */
export const matchExistInArrays = (array1: unknown[], array2: unknown[]): boolean => {
    return array1.some(item => array2.includes(item));
};

/**
 * Finds and returns elements from the parentArray that do not exist in the childArray.
 *
 * @param {unknown[]} parentArray - The source array to check.
 * @param {unknown[]} childArray - The reference array to compare against.
 * @returns {any[]} - An array containing elements from `parentArray` that are not present in `childArray`.
 */
export const findUniqueItemsInParentArray = (parentArray: unknown[], childArray: unknown[]): any[] => {
    const childSet = new Set(childArray);
    return parentArray.filter(item => !childSet.has(item));
};
