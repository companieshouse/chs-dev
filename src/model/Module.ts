/**
 * Represents a collection of inter-related services. Potentially around
 * journey or function
 */
export interface Module {

    /**
     * Name of the module - i.e. the directory name
     */
    name: string;
}

export default Module;
