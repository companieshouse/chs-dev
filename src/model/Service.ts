import { DependencyNode } from "./DependencyGraph.js";
import { OrNull } from "./OrNull.js";

/**
 * Represents a service within the development environment. A service is a
 * single deployable unit - typically container.
 */
export interface Service {
    /**
     * Meaningful name for service, unique within the environment
     */
    name: string;

    /**
     * Name of the module which the service resides
     */
    module: string;

    /**
     * Human readable description of the service
     */
    description?: OrNull<string>;

    /**
     * Path to the Docker Compose file for the service
     */
    source: string;

    /**
     * List of services names (exhaustive) that the service depends on.
     *
     * This will include all of the direct and indirect/transitive dependencies
     * the service has.
     */
    dependsOn: string[];

    /**
     * Tree of the dependencies for the service.
     *
     * This will include all of the direct and indirect/transitive dependencies
     * the service has.
     */
    dependencyTree?: DependencyNode;

    /**
     * The number of times this service is used by other services.
     *
     * This is the number of services that have this service in their `dependsOn` list and
     * is not the same as `numberOfDependencies` which counts the number of
     * services that the service depends on.
     */
    timesUsedByOtherServices?: number;
    /**
     * The number of services that depend on this service.
     *
     * This is the number of services that this service depends on either directly or
     * indirectly. This is not the same as `timesUsedByOtherServices` which counts
     * the number of time a service appears on the `dependsOn` list.
     */
    numberOfDependencies?: number;
    /**
     * Details about the code repository for the service (required for
     * development mode)
     */
    repository: OrNull<{ url: string; branch?: string | null; }>;

    /**
     * When supplied will specify a common Dockerfile to use to build the
     * service for development mode
     */
    builder: string;

    /**
     * Other pertinent information about the service which may vary
     * service to service.
     */
    metadata: Record<string, OrNull<string> | undefined>;
}

export default Service;
