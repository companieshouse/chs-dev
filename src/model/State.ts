/**
 * Represents the state of the world for the environment
 */
export interface State {

    /**
     * list of module names of the modules the user has enabled
     */
    modules: string[];

    /**
     * list of the service names the user has enabled
     */
    services: string[];

    /**
     * list of the service names which are within development mode within the
     * environment
     */
    servicesWithLiveUpdate: string[];

    /**
     * list of service names which are removed from the deployed environment
     */
    excludedServices: string[];
}

export default State;
