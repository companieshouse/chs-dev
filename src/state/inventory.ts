import { existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "fs";
import { basename, join } from "path";

import yaml from "yaml";
import glob from "glob";

import { createHash } from "crypto";
import { Service } from "../model/Service.js";
import { Module } from "../model/Module.js";
import { readServices } from "./service-reader.js";
import { DependencyNameResolver } from "./dependency-name-resolver.js";

interface InventoryCache {
  hash: string;
  services: Service[];
  modules: Module[]
}

export class Inventory {
    readonly inventoryCacheFile: string;
    private __serviceFiles: string[] | undefined;

    constructor (private path: string, cacheDir: string) {
        this.path = path;
        this.inventoryCacheFile = join(cacheDir, `${basename(path)}.inventory.yaml`);

        if (!existsSync(cacheDir)) {
            mkdirSync(cacheDir, {
                recursive: true
            });
        }
    }

    get modules (): Module[] {
        return this.getFromCache(inventoryCache => inventoryCache.modules);
    }

    get services (): Service[] {
        return this.getFromCache(inventoryCache => inventoryCache.services);
    }

    private getFromCache<T> (cacheSupplier: (inventoryCache: InventoryCache) => T): T {
        const inventoryCache = this.inventoryCache;

        if (typeof inventoryCache !== "undefined") {
            if (inventoryCache.hash === this.hashServiceFiles()) {
                return cacheSupplier(inventoryCache);
            }
        }

        this.updateCache();

        return this.getFromCache(cacheSupplier);
    }

    private updateCache () {
        const inventory = {
            hash: this.hashServiceFiles(),
            services: this.loadServices(),
            modules: this.loadModules()
        };

        writeFileSync(this.inventoryCacheFile, yaml.stringify(inventory));
    }

    private loadModules (): Module[] {
        return readdirSync(join(this.path, "services", "modules"), { withFileTypes: true })
            .filter(item => item.isDirectory())
            .map(item => {
                return {
                    name: item.name
                };
            });
    }

    private loadServices (): Service[] {
        const partialServices: Partial<Service>[] = this.serviceFiles.flatMap(readServices);

        const dependencyNameResolver = new DependencyNameResolver(partialServices);

        return partialServices.map(service => ({
            ...service,
            dependsOn: dependencyNameResolver.fullDependencyListIncludingTransitive(service.dependsOn as string[])
        } as Service));
    }

    private hashServiceFiles () {
        const sha256Hash = createHash("sha256");

        this.serviceFiles
            .map(serviceFile => readFileSync(serviceFile))
            .forEach((serviceFile:any) => {
                sha256Hash.update(serviceFile);
            });

        return sha256Hash.digest("hex");
    }

    private get inventoryCache (): InventoryCache | undefined {
        if (existsSync(this.inventoryCacheFile)) {
            return yaml.parse(readFileSync(this.inventoryCacheFile).toString("utf-8"));
        }

        return undefined;
    }

    private get serviceFiles (): string[] {
        if (typeof this.__serviceFiles === "undefined") {
            this.__serviceFiles = glob.sync(`${this.path}/services/**/*.docker-compose.yaml`);
        }
        return this.__serviceFiles;
    }
}
