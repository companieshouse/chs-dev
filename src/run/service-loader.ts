import { deduplicate } from "../helpers/array-reducers.js";
import Service from "../model/Service.js";
import State from "../model/State.js";
import { Inventory } from "../state/inventory.js";

type LoadedService = Service & {
  liveUpdate: boolean;
}

export class ServiceLoader {

    private readonly inventory: Inventory;

    constructor (inventory: Inventory) {
        this.inventory = inventory;
    }

    loadServices (state: State): LoadedService[] {
        const withLiveUpdate = (service: Service) => ({
            ...service,
            liveUpdate: state.servicesWithLiveUpdate.includes(service.name)
        });

        const loadedServices = this.inventory.services
            .filter(service => state.services.includes(service.name) || state.modules.includes(service.module))
            .map(withLiveUpdate);

        const fullDependencyList = loadedServices
            .flatMap(service => service.dependsOn)
            .reduce(deduplicate, [])
            .filter(serviceName => !loadedServices.some(service => service.name === serviceName))
            .map(serviceName => this.findService(serviceName) as Service)
            .map(withLiveUpdate);

        return [
            ...loadedServices,
            ...fullDependencyList
        ] as LoadedService[];
    }

    private findService (serviceName: string): Service | undefined {
        return this.inventory.services.find(service => service.name === serviceName);
    }

}
