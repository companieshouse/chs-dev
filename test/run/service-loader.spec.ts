import { expect, jest } from "@jest/globals";
import Module from "../../src/model/Module";
import Service from "../../src/model/Service";
import State from "../../src/model/State";
import { deduplicate } from "../../src/helpers/array-reducers";
import { Inventory } from "../../src/state/inventory";
import { ServiceLoader } from "../../src/run/service-loader";

const serviceOne = {
    name: "service-one",
    module: "module-one",
    source: "./module-one/service-one.docker-compose.yaml",
    dependsOn: [],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceTwoTilt = {
    name: "service-two",
    module: "module-one",
    source: "./module-one/tilt/service-two.docker-compose.yaml",
    dependsOn: ["service-one"],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceTwo = {
    name: "service-two",
    module: "module-one",
    source: "./module-one/service-two.docker-compose.yaml",
    dependsOn: ["service-one"],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceThree = {
    name: "service-three",
    module: "module-one",
    source: "./module-one/service-three.docker-compose.yaml",
    dependsOn: ["service-one"],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceFour = {
    name: "service-four",
    module: "module-two",
    source: "./module-two/service-four.docker-compose.yaml",
    dependsOn: ["service-one", "service-two", "service-three"],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceFive = {
    name: "service-five",
    module: "module-two",
    source: "./module-two/service-five.docker-compose.yaml",
    dependsOn: ["service-one", "service-two", "service-three", "service-four"],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceSix = {
    name: "service-six",
    module: "module-two",
    source: "./module-two/service-six.docker-compose.yaml",
    dependsOn: [
        "service-one",
        "service-two",
        "service-three",
        "service-four",
        "service-five"
    ],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceSeven = {
    name: "service-seven",
    module: "module-three",
    source: "./module-three/service-seven.docker-compose.yaml",
    dependsOn: [
        "service-one",
        "service-two",
        "service-three",
        "service-four",
        "service-five",
        "service-six"
    ],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceEight = {
    name: "service-eight",
    module: "module-three",
    source: "./module-three/service-eight.docker-compose.yaml",
    dependsOn: [
        "service-one",
        "service-two",
        "service-three",
        "service-four",
        "service-five",
        "service-six",
        "service-seven"
    ],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceNine = {
    name: "service-nine",
    module: "module-three",
    source: "./module-three/service-seven.docker-compose.yaml",
    dependsOn: [
        "service-one",
        "service-two",
        "service-three",
        "service-four",
        "service-five",
        "service-six",
        "service-seven",
        "service-eight",
        "service-eleven"
    ],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceTen = {
    name: "service-ten",
    module: "module-four",
    source: "./module-four/service-ten.docker-compose.yaml",
    dependsOn: [],
    repository: null,
    builder: "",
    metadata: {}
};

const serviceTenTilt = {
    name: "service-ten",
    module: "module-four",
    source: "./module-four/tilt/service-ten.docker-compose.yaml",
    dependsOn: [],
    repository: null,
    builder: "",
    metadata: {}
};

const services: Service[] = [
    serviceOne,
    serviceTwoTilt,
    serviceTwo,
    serviceThree,
    serviceFour,
    serviceFive,
    serviceSix,
    serviceSeven,
    serviceEight,
    serviceNine,
    serviceTen,
    serviceTenTilt
];

const modules: Module[] = services.map(service => service.module)
    .reduce(deduplicate, [])
    .map(modName => ({ name: modName }));

const inventory = {
    services,
    modules
} as Inventory;

describe("ServiceLoader", () => {

    const serviceLoader = new ServiceLoader(inventory);
    let state: State;

    beforeEach(() => {
        jest.resetAllMocks();
    });

    it("loads services correctly when all services enabled, none in dev mode", () => {
        state = {
            services: [
                "service-three",
                "service-two"
            ],
            modules: [],
            servicesWithLiveUpdate: [],
            excludedServices: []
        };

        const result = serviceLoader.loadServices(state);

        expect(result).toMatchSnapshot();
    });

    it("loads modules, no services enabled, none in dev mode", () => {
        state = {
            services: [],
            modules: ["module-three"],
            servicesWithLiveUpdate: [],
            excludedServices: []
        };

        const result = serviceLoader.loadServices(state);

        expect(result).toMatchSnapshot();
    });

    it("loads module and service, none in dev mode", () => {
        state = {
            services: ["service-four"],
            modules: ["module-four"],
            servicesWithLiveUpdate: [],
            excludedServices: []
        };

        const result = serviceLoader.loadServices(state);

        expect(result).toMatchSnapshot();
    });

    it("loads services in dev mode", () => {
        state = {
            services: ["service-four"],
            modules: ["module-four"],
            servicesWithLiveUpdate: ["service-four", "service-ten"],
            excludedServices: []
        };

        const result = serviceLoader.loadServices(state);

        expect(result).toMatchSnapshot();
    });

    it("does not load a transient service in dev mode", async () => {
        state = {
            services: [],
            modules: ["module-three"],
            servicesWithLiveUpdate: ["service-four"],
            excludedServices: []
        };

        const result = serviceLoader.loadServices(state);

        expect(result).toMatchSnapshot();

    });
});
