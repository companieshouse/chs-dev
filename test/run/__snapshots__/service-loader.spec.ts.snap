// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceLoader does not load a transient service in dev mode 1`] = `
[
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
      "service-five",
      "service-six",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-three",
    "name": "service-seven",
    "repository": null,
    "source": "./module-three/service-seven.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
      "service-five",
      "service-six",
      "service-seven",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-three",
    "name": "service-eight",
    "repository": null,
    "source": "./module-three/service-eight.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
      "service-five",
      "service-six",
      "service-seven",
      "service-eight",
      "service-eleven",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-three",
    "name": "service-nine",
    "repository": null,
    "source": "./module-three/service-seven.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-one",
    "repository": null,
    "source": "./module-one/service-one.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-two",
    "repository": null,
    "source": "./module-one/service-two.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-three",
    "repository": null,
    "source": "./module-one/service-three.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-two",
    "name": "service-four",
    "repository": null,
    "source": "./module-two/service-four.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-two",
    "name": "service-five",
    "repository": null,
    "source": "./module-two/service-five.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
      "service-five",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-two",
    "name": "service-six",
    "repository": null,
    "source": "./module-two/service-six.docker-compose.yaml",
  },
]
`;

exports[`ServiceLoader loads module and service, none in dev mode 1`] = `
[
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-two",
    "name": "service-four",
    "repository": null,
    "source": "./module-two/service-four.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-four",
    "name": "service-ten",
    "repository": null,
    "source": "./module-four/service-ten.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-one",
    "repository": null,
    "source": "./module-one/service-one.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-two",
    "repository": null,
    "source": "./module-one/service-two.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-three",
    "repository": null,
    "source": "./module-one/service-three.docker-compose.yaml",
  },
]
`;

exports[`ServiceLoader loads modules, no services enabled, none in dev mode 1`] = `
[
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
      "service-five",
      "service-six",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-three",
    "name": "service-seven",
    "repository": null,
    "source": "./module-three/service-seven.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
      "service-five",
      "service-six",
      "service-seven",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-three",
    "name": "service-eight",
    "repository": null,
    "source": "./module-three/service-eight.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
      "service-five",
      "service-six",
      "service-seven",
      "service-eight",
      "service-eleven",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-three",
    "name": "service-nine",
    "repository": null,
    "source": "./module-three/service-seven.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-one",
    "repository": null,
    "source": "./module-one/service-one.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-two",
    "repository": null,
    "source": "./module-one/service-two.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-three",
    "repository": null,
    "source": "./module-one/service-three.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-two",
    "name": "service-four",
    "repository": null,
    "source": "./module-two/service-four.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-two",
    "name": "service-five",
    "repository": null,
    "source": "./module-two/service-five.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
      "service-four",
      "service-five",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-two",
    "name": "service-six",
    "repository": null,
    "source": "./module-two/service-six.docker-compose.yaml",
  },
]
`;

exports[`ServiceLoader loads services correctly when all services enabled, none in dev mode 1`] = `
[
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-two",
    "repository": null,
    "source": "./module-one/service-two.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-three",
    "repository": null,
    "source": "./module-one/service-three.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-one",
    "repository": null,
    "source": "./module-one/service-one.docker-compose.yaml",
  },
]
`;

exports[`ServiceLoader loads services in dev mode 1`] = `
[
  {
    "builder": "",
    "dependsOn": [
      "service-one",
      "service-two",
      "service-three",
    ],
    "liveUpdate": true,
    "metadata": {},
    "module": "module-two",
    "name": "service-four",
    "repository": null,
    "source": "./module-two/service-four.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [],
    "liveUpdate": true,
    "metadata": {},
    "module": "module-four",
    "name": "service-ten",
    "repository": null,
    "source": "./module-four/service-ten.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-one",
    "repository": null,
    "source": "./module-one/service-one.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-two",
    "repository": null,
    "source": "./module-one/service-two.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "service-one",
    ],
    "liveUpdate": false,
    "metadata": {},
    "module": "module-one",
    "name": "service-three",
    "repository": null,
    "source": "./module-one/service-three.docker-compose.yaml",
  },
]
`;
